shipctl/x86_64/WindowsServer_2016/utility.ps1

```diff
@@ -540,3 +540,197 @@ function get_git_changes() {
+      }
+    }
+  }
+
+  $env:opt_username = $username
+  if (!$env:opt_username) {
+    $opt_username = $env:NOTIFY_USERNAME
+    if (!$env:opt_username) {
+      $env:opt_username = "Shippable"
+    }
+  }
+  $env:opt_payload = $payload
+  if (!$env:opt_payload) {
+      $env:opt_payload = $env:NOTIFY_PAYLOAD
+      if (!$env:opt_payload) {
+        $env:opt_payload = ""
+      }
+  }
+
+  $default_slack_payload = '{"username":"$opt_username","attachments":[{"pretext":"$opt_pretext","text":"$opt_text","color":"$opt_color"}],"channel":"$opt_recipient","icon_url":"$opt_icon_url"}'
+  $default_webhook_payload = '{"username":"$opt_username","pretext":"$opt_pretext","text":"$opt_text","color":"$opt_color","channel":"$opt_recipient","icon_url":"$opt_icon_url"}'
+  $default_payload=""
+
+  if ($r_mastername -eq "Slack" -or $r_mastername -eq "slackKey") {
+    $default_payload=$default_slack_payload
+    $meta=$(get_resource_meta "$env:opt_resource")
+    $recipients_list = $(get_resource_version_key "$env:opt_resource" "recipients")
+  } elseif ($r_mastername -eq "webhook" -or $r_mastername -eq "webhookV2") {
+    $r_authorization=$(get_integration_resource_field "$env:opt_resource" authorization)
+    if (!$r_authorization) {
+      $r_authorization = ""
+    }
+    $default_payload=$default_webhook_payload
+  } else {
+    throw "Error: unsupported notification type: $r_mastername"
+  }
+  if ($env:opt_payload) {
+    if (!(Test-Path $env:opt_payload)) {
+      throw "Error: file not found at path: $opt_payload"
+    }
+    try {
+      $result = Get-Content $env:opt_payload -Raw | ConvertFrom-Json -ErrorAction Stop
+    } catch {
+      throw "Error: payload is not valid JSON"
+    }
+
+    Write-Output "sending notification"
+    _send_notification -payload "$env:opt_payload" -auth "$r_authorization" -endpoint "$r_endpoint"
+
+  } else {
+    if ($($recipients_list.count) -gt 0 -and !$env:opt_recipient) {
+      foreach ($recipient in $recipients_list) {
+        $env:opt_recipient = $recipient
+
+        Out-File -FilePath "$env:TEMP/payload.json" -InputObject $default_payload
+        $env:opt_payload = "$env:TEMP/payload.json"
+        shipctl replace $env:opt_payload
+
+        try {
+          $result = Get-Content $env:opt_payload -Raw | ConvertFrom-Json -ErrorAction Stop
+        } catch {
+          throw "Error: payload is not valid JSON"
+        }
+
+        Write-Output "sending notification to $env:opt_recipient"
+        _send_notification -payload "$env:opt_payload" -auth "$r_authorization" -endpoint "$r_endpoint"
+      }
+    } else {
+
+      Out-File -FilePath "$env:TEMP/payload.json" -InputObject $default_payload
+      $env:opt_payload="$env:TEMP/payload.json"
+      shipctl replace $env:opt_payload
+
+      $fileContent = Get-Content "$env:opt_payload"
+      # Support ${NAME} and $NAME, in that order of preference
+      $fileContent = $fileContent.Replace("$" + "opt_recipient", "")
```

Comment:
some info for context here:
- the windows version of shipctl replace will NOT replace variables that are not defined in the environment (while the bash version replaces them with empty string "")
- this step only occurs when the user has not provided their own payload, and has not defined a recipient. Since the default payload is pre-defined, it is guaranteed then that $opt_recipient will be present in the file still.
- the comment that says ${name} is wrong, i should have deleted that.
- the reason i took this approach of replacing the file is that i could not find any way in powershell to define a variable as an empty string, such that a shipctl replace by itself would actually replace it with a `""`.  I searched online and tried a few different approaches, but this direct replace made sense to me in this case.

Reply:
Ok, that makes sense. I wasn't aware that `$env:foo = ""` doesn't create an env var with an empty value. It actually _unsets_ the var, which is weird.

And, yeah, replace on Windows doesn't use envsubst (because it doesn't exist for Windows), so we had to roll our own.